internal/controller/serving/inferenceservice_controller_test.go

```diff
@@ -1289,6 +1294,567 @@ var _ = Describe("InferenceService Controller", func() {
+			kedaReconciler = reconcilers.NewKServeKEDAReconciler(k8sClient)
+		})
+
+		Context("when InferenceServices are configured to scale on KEDA metrics", func() {
+			var isvc *kservev1beta1.InferenceService
+			var isvc2 *kservev1beta1.InferenceService
+
+			const fake = "fake" // for the love of the linter :)
+
+			BeforeEach(func() {
+				isvc = makeKedaTestISVC(testNs, names.SimpleNameGenerator.GenerateName("keda-isvc"), true)
+				Expect(k8sClient.Create(ctx, isvc)).Should(Succeed())
+				Expect(k8sClient.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace}, isvc)).Should(Succeed())
+
+				isvc2 = makeKedaTestISVC(testNs, names.SimpleNameGenerator.GenerateName("keda-isvc-2"), true)
+				Expect(k8sClient.Create(ctx, isvc2)).Should(Succeed())
+				Expect(k8sClient.Get(ctx, types.NamespacedName{Name: isvc2.Name, Namespace: isvc2.Namespace}, isvc2)).Should(Succeed())
+
+				// Safety checks
+				Expect(isvc.Spec.Predictor.AutoScaling).ToNot(BeNil())
+				Expect(isvc2.Spec.Predictor.AutoScaling).ToNot(BeNil())
+				Expect(isvc.ObjectMeta.UID).ToNot(BeEmpty())
+				Expect(isvc2.ObjectMeta.UID).ToNot(BeEmpty())
+
+				err := kedaReconciler.Reconcile(ctx, GinkgoLogr, isvc)
+				Expect(err).NotTo(HaveOccurred())
+				err = kedaReconciler.Reconcile(ctx, GinkgoLogr, isvc2)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("Must have owner reference to InferenceService", func() {
+				for kind, obj := range getAllKedaTestResources(ctx, testNs) {
+					Expect(obj).To(Not(BeNil()), fmt.Sprintf("Obj: %#v", obj))
+					Expect(hasOwnerReferenceByUID(obj, isvc.UID)).To(BeTrue(), fmt.Sprintf(`
+%s should be owned by ISVC
+obj: %#v
+isvc: %#v
+`, kind, obj, isvc))
+					Expect(hasOwnerReferenceByUID(obj, isvc2.UID)).To(BeTrue(), fmt.Sprintf(`
+%s should be owned by ISVC
+obj: %#v
+isvc: %#v
+`, kind, obj, isvc2))
+				}
+			})
+
+			It("Must remove owner reference from deleted InferenceService", func() {
+				Expect(k8sClient.Delete(ctx, isvc2)).Should(Succeed())
+				Expect(kedaReconciler.Delete(ctx, GinkgoLogr, isvc2)).To(Succeed())
+
+				for kind, obj := range getAllKedaTestResources(ctx, testNs) {
+					Expect(obj).To(Not(BeNil()), fmt.Sprintf("Obj: %#v", obj))
+					Expect(hasOwnerReferenceByUID(obj, isvc.UID)).To(BeTrue(), fmt.Sprintf(`
+%s should be owned by ISVC
+obj: %#v
+isvc: %#v
+`, kind, obj, isvc))
+					Expect(hasOwnerReferenceByUID(obj, isvc2.UID)).To(BeFalse(), fmt.Sprintf(`
+%s should not be owned by ISVC
+obj: %#v
+isvc: %#v
+`, kind, obj, isvc2))
+				}
+			})
+
+			It("Must update managed role when it diverges", func() {
+				role := &rbacv1.Role{}
+				Expect(k8sClient.Get(ctx, types.NamespacedName{Namespace: isvc.Namespace, Name: reconcilers.KEDAPrometheusAuthMetricsReaderRoleName}, role)).Should(Succeed())
+				role.Rules[0].Resources = []string{fake}
+				Expect(k8sClient.Update(ctx, role)).To(Succeed())
+
+				Expect(kedaReconciler.Reconcile(ctx, GinkgoLogr, isvc)).To(Succeed())
```

Comment:
I see that this pattern of directly calling Reconcile is used in several places. My impression is that while it helps tests run quickly and deterministically, it might not catch potential conflicts or race conditions that can occur when reconciles run concurrently or interact asynchronously with other controllers.

Given that, I wonder if we might be missing some concurrency or interaction issues that only show up in a more realistic environment. What do you think about complementing these tests with some integration-style tests that run the manager and allow reconciles to trigger automatically? That way, we might be able to catch those subtle race conditions or conflicts better.

Would love to hear your thoughts on this approach.